### PR TITLE
Inference on layer

### PR DESCRIPTION
**Pull request by C-Achard**
Completed PR for #7 

Features :
- Working option to run inference on a single layer, with all functionalities
- Improvements to inference code (worker and widget)
- Fix for deadlock in worker thread on exception (#25)

ISSUES :
- ~~Proper button behaviour~~
- Orientation of the output can be wrong sometimes, will have to determine best default case -> will be addressed separately
- ~~Testing all features~~
- ~~Proper error handling/reset when running a job without a folder/layer selected~~

### CODE REVIEW

File: napari_cellseg3d/model_workers.py [hunk 1]
```diff
@@ -289,11 +299,330 @@ def log_parameters(self):
...
+            )  # TODO test with area mesh function
+            return data_dict
+
+        # except ValueError as e:
+        #     self.log(f"Error occurred during stats computing : {e}")
+        #     return None
+        else:
+            return None
+
+    def inference_on_layer(self, image, model, post_process_transforms):
```
Comment (vidalmaxime) [inline napari_cellseg3d/model_workers.py:553]: Seems to be a bunch of shared code with inference_on_list, could you call a new shared function from both of these ?

File: napari_cellseg3d/model_workers.py [hunk 2]
```diff
@@ -289,11 +299,330 @@ def log_parameters(self):
...
+        #     out = anisotropic_transform(out[0])
+        ##################
+        ##################
+        ##################
+        out = post_process_transforms(out)
+        out = np.array(out).astype(np.float32)
+        out = np.squeeze(out)
+        to_instance = out  # avoid post processing since thresholding is done there anyway
+
+        # batch_len = out.shape[1]
```
Comment (vidalmaxime) [inline napari_cellseg3d/model_workers.py:577]: Remove comments ?

File: napari_cellseg3d/model_workers.py [hunk 3]
```diff
@@ -334,261 +663,146 @@ def inference(self):
             torch.set_num_threads(1)  # required for threading on macOS ?
             self.log("Number of threads has been set to 1 for macOS")
 
-        images_dict = self.create_inference_dict(self.images_filepaths)
+        # images_dict = self.create_inference_dict(self.images_filepaths)
```
Comment (vidalmaxime) [inline napari_cellseg3d/model_workers.py:666]: Remove comments ?

File: napari_cellseg3d/model_instance_seg.py [hunk 4]
```diff
@@ -194,15 +194,19 @@ def volume_stats(volume_image):
...
+    # sphericity_va = []
+    def sphericity(region):
+        try:
+            return sphericity_axis(
+                region.axis_major_length * 0.5, region.axis_minor_length * 0.5
+            )
+        except ValueError:
+            return (
+                np.nan
+            )  # FIXME better way ? inconsistent errors in region.axis_minor_length
```
[the remark this answers is not in the document]
Comment (C-Achard) [reply napari_cellseg3d/model_instance_seg.py:207]: The source is region.axis_minor_length from skimage regionprops I think, it crashes sometimes, might be due to concavity of volumes as we discussed. I haven't searched deeper

**Review by MMathisLab — APPROVED** [review]
agree with removing some print statements, looks great!

File: napari_cellseg3d/model_workers.py (hunk 1 above)
Comment (C-Achard) [reply napari_cellseg3d/model_workers.py:618]: Fixed in 8cf755c

**Review by vidalmaxime — APPROVED** [review]

**Pull request merged**